pkg/codegen/schema/schema.go

```diff
@@ -0,0 +1,898 @@
+type DefaultValue struct {
+	// Value specifies a static default value, if any. This value must be representable in the Pulumi schema type
+	// system, and its type must be assignable to that of the property to which the default applies.
+	Value interface{}
+	// Environment specifies a set of environment variables to probe for a default value.
+	Environment []string
+	// Language specifies additional language-specific data about the default value.
+	Language map[string]json.RawMessage
+}
+
+// Property describes an object or resource property.
+type Property struct {
+	// Name is the name of the property.
+	Name string
+	// Comment is the description of the property, if any.
+	Comment string
+	// Type is the type of the property.
+	Type Type
+	// DefaultValue is the default value for the property, if any.
+	DefaultValue *DefaultValue
+	// IsRequired is true if the property must always be populated.
+	IsRequired bool
+	// Language specifies additional language-specific data about the property.
+	Language map[string]json.RawMessage
+}
+
+// Alias describes an alias for a Pulumi resource.
+type Alias struct {
+	// Name is the "name" portion of the alias, if any.
+	Name *string
+	// Project is the "project" portion of the alias, if any.
+	Project *string
+	// Type is the "type" portion of the alias, if any.
+	Type *string
+}
+
+// Resource describes a Pulumi resource.
+type Resource struct {
+	// Token is the resource's Pulumi type token.
+	Token string
+	// Comment is the description of the resource, if any.
+	Comment string
+	// IsProvider is true if the resource is a provider resource.
+	IsProvider bool
+	// InputProperties is the list of the resource's input properties.
+	InputProperties []*Property
+	// Properties is the list of the resource's output properties. This should be a superset of the input properties.
+	Properties []*Property
+	// StateInputs is the set of inputs used to get an existing resource, if any.
+	StateInputs *ObjectType
+	// Aliases is the list of aliases for the resource.
+	Aliases []*Alias
+	// DeprecationMessage indicates whether or not the resource is deprecated.
+	DeprecationMessage string
+	// Language specifies additional language-specific data about the resource.
+	Language map[string]json.RawMessage
+}
+
+// Function describes a Pulumi function.
+type Function struct {
+	// Token is the function's Pulumi type token.
+	Token string
+	// Comment is the description of the function, if any.
+	Comment string
+	// Inputs is the bag of input values for the function, if any.
+	Inputs *ObjectType
+	// Outputs is the bag of output values for the function, if any.
+	Outputs *ObjectType
+	// DeprecationMessage indicates whether or not the function is deprecated.
+	DeprecationMessage string
+	// Language specifies additional language-specific data about the function.
+	Language map[string]json.RawMessage
+}
+
+// Package describes a Pulumi package.
+type Package struct {
+	moduleFormat *regexp.Regexp
```

Comment:
interesting.  is this just for validation on our part?

Reply:
This is to accommodate differences in module naming between different providers.

The Terraform-derived providers use this format for member tokens:
```
pkg:namespace/member:member
```

The k8s provider uses this format:
```
pkg:namespace:member
```

The latter format is the format I'd like to standardize on, but doing so would require aliasing the type portion of every TF-derived resource. The regex-based approach allows us to avoid that, which seems nice.

Reply:
This is to accommodate differences in the format of the module portion of member tokens. Ideally, all of our providers would use the following format:
```
pkg:namespace:member
```
Where `namespace` is a slash-separated path. The k8s provider uses this format today, e.g.:
```
kubernetes:meta/v1:ObjectMeta
```

Unfortunately, the TF providers do something a bit different, and format the type token like so:
```
pkg:namespace/member:member
```

Using a regex to extract the namespace allows us to accommodate this difference without aliasing every resource defined by a TF provider.